### PR TITLE
PWA

### LINKED ISSUE #1666: Storefront as PWA
Since Saleor delivers both RWD and Payment Request (well, almost), I think it is worth considering making it full PWA (https://developers.google.com/web/progressive-web-apps/).

The main problem is offline mode and I believe it would be resolved only when we rewrite whole front-end as single page application in React. It would also require GraphQL API.
I'd like to propose working on making Saleor PWA in two stages:

Online only
- 
The main goal in this phase would be optimizing current Saleor storefront in terms of performance - as latest Lighthouse audit shows, there are some things which we can improve:
zrzut ekranu 2018-01-23 o 17 42 00
I think that it is worth to think about adding image lazy-loaders and optimizing assets' (`.css`, `.js`) size. Also please look at Webpack's code splitting (https://webpack.js.org/guides/code-splitting/) and share your thoughts about it, beacuse I really can't decide whether Saleor could use it or it's not worth it.
Besides pure performance and RWD, application is considered as PWA when it can be "installed" on the device, providing custom splash screen, setting theme colors and caching assets (and maybe some portion of data) in background. It would require serving `manifest.json` and providing service worker, which acts like proxy in offline mode, serving custom connection error page.

Online + Offline
-
In this stage Saleor goes full PWA, getting whole storefront rewritten as single-page application in React, with fully operational API. Going PWA's direction requires moving all filtering / sorting logic to be moved from Python to JS, beacuse of caching reasons. Offline mode should (at least I think) provide interactions with product list, like viewing, sorting, filtering and searching for a specific product. Or maybe someone has better idea?

EDIT actually it is possible to mock GraphQL API by service worker. Maybe this isn't bad idea to cache all products by storing them in IndexedDB, and do all the filtering / sorting things purely by JS in offline mode?

Feel free to share your opinion.

### PR DESCRIPTION
**Pull request by dominik-zeglen**
Ref #1666 

I'd be cool if someone more from backend team would look at it and refactor serviceworker and manifest serving.

### CODE REVIEW

File: saleor/static/js/components/misc.js [hunk 1]
```diff
@@ -25,4 +26,15 @@ export default $(document).ready((e) => {
 
   let hash = window.location.hash;
   $('.nav-tabs a[href="' + hash + '"]').tab('show');
+
+  if ('serviceWorker' in navigator) {
+    navigator.serviceWorker
+      .register('/serviceworker.js')
+      .then((registration) => {
+        console.log('Registration successful, scope is:', registration.scope);
```
Comment (elwoodxblues) [inline saleor/static/js/components/misc.js]: This looks like debug (and the log below).

File: templates/manifest.json [hunk 2]
```diff
@@ -0,0 +1,63 @@
+{
+  "lang": "en",
+  "dir": "ltr",
+  "name": "Saleor demo shop",
+  "description": "Saleor is best storefront engine evar",
```
Comment (elwoodxblues) [inline templates/manifest.json]: This doesn't look very serious.

File: templates/manifest.json [hunk 3]
```diff
@@ -0,0 +1,63 @@
+{
+  "lang": "en",
+  "dir": "ltr",
+  "name": "Saleor demo shop",
```
Comment (elwoodxblues) [inline templates/manifest.json]: Simply "Saleor".

File: templates/manifest.json [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  "description": "Saleor is best storefront engine evar",
+  "short_name": "Saleor Demo",
```
Comment (elwoodxblues) [inline templates/manifest.json]: Here as well I guess "Saleor" should be okay.

File: templates/manifest.json [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+  "icons": [
+    {
+      "src": "/static/assets/favicons/android-chrome-36x36.png",
```
Comment (elwoodxblues) [inline templates/manifest.json]: Those paths may be different depending on the environment. I think this whole file should be generated by Django.

File: templates/serviceworker.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+          .catch(e => console.error(e));
+      })
+      .catch(e => console.error(e))
+  );
+});
+
+// when the browser fetches a url
+self.addEventListener('fetch', (event) => {
+  if (event.request.url.match(/http:\/\/.*((\.(css|js|svg)$)|(placeholder\d+x\d+\.png$))/)) {
+    console.log(event.request.url);
```
Comment (elwoodxblues) [inline templates/serviceworker.js]: Please remove this and other occurrences of console.log and any other debug code.

**Comment by patrys** [comment]
Why are we randomly adding offline service workers?

File: saleor/core/urls.py [hunk 7]
```diff
@@ -10,5 +11,12 @@
         name='impersonate-start'),
     url(r'^impersonate/stop/$', views.stop_impersonate,
         name='impersonate-stop'),
-    url(r'^404', views.handle_404, name='handle-404')
+    url(r'^404', views.handle_404, name='handle-404'),
+    url(r'^offline$', views.offline, name='offline'),
+    url(r'^manifest\.json$', TemplateView.as_view(
+        template_name='manifest.json', content_type='application/json')),
+    url(r'^serviceworker\.js', TemplateView.as_view(
+        template_name="serviceworker.js",
```
Comment (patrys) [inline saleor/core/urls.py]: Interesting engineering choices 🐼 

mechanic

**Comment by elwoodxblues** [comment]
If there are no further comments on this PR, I'd like to merge it.

**Comment by Pacu2** [comment]
Please add `lazypreload` class to all lazyloaded images, it will load them after page loads. Otherwise, if someone scrolls fast he will be smashed by multiple thumbnails.
Up for a discussion, if we should use a placeholder when image is being lazyloaded.

**Comment by elwoodxblues** [comment]
CC @dominik-zeglen ⬆️

**Comment by dominik-zeglen** [comment]
@Pacu2 big :+1:

File: templates/manifest.json [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    },
+    {
+      "src": "{% webpack_static "favicons/android-chrome-512x512.png" %}",
+      "sizes": "512x512",
+      "type": "image/png"
+    }
+  ],
+  "scope": "/",
+  "start_url": "{% url 'home' %}",
+  "display": "fullscreen",
```
Comment (Pacu2) [inline templates/manifest.json]: Is this a valid value? Shouldn't it be `standalone`?

Comment (dominik-zeglen) [reply templates/manifest.json]: > fullscreen: All of the available display area is used and no user agent chrome is shown. 

It also fallbacks to `standalone` if `fullscreen` cannot be used.

Source: https://developer.mozilla.org/en-US/docs/Web/Manifest

**Pull request merged**